docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,126 @@
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. @timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
+- Runtimes (only one of these, in addition to Common, should be specified for a given Kubelet)
+    + Common
+    + Rkt
+    + Docker
+- Events
+- Security/Policies
+- Images (might be better as a runtime sub-category)
+- Resource Management
+    + GC/Eviction
+    + Allocatable/SystemReserved/KubeReserved
+    + Custom resources? e.g. NvidiaGPUs?
+- Cloud-provider-specific settings
+- Cluster/API
+    + KubeAPIQPS
+    + ContentType
+- SystemSpecification (this might be a non-category, all of the available hardware should theoretically be detectable by the Kubelet)
+    + NodeIP (probably belongs in networking)
+    + NvidiaGPUs could be here, but shouldn't the Kubelet be able to detect this?
+- Experimental/Volatile Features (could be tricky, might be better off using alpha versioning to separate experiments from the stable API)
+
+---
+
+### Bringing up a master node
+
+- Start the Kubelet with `minimal-config`
+- On-disk static pod manifests are used to bring up the API server, scheduler, controller-manager, etc.
+Once the master is ready, the user creates some configs via the API:
+- They might create a `global-kubelet-configuration`
+- They might create a `<node-name>-kubelet-configuration`, e.g. if they want a master-specific configuration.
+- Then the Kubelets' configuration sync loops will see the new config, and the Kubelets will restart to take up that new config.
+
+### Bringing up a Kubelet
+
+- Start the Kubelet with `minimal-config`
+- If `<current-node-name>-kubelet-config` exists on the API server, download and use the `<current-node-name>-kubelet-config`
+- else if `global-kubelet-config` exists on the API server, download and use the `global-kubelet-config`.
+- else if e.g. an `on-disk-config` flag is set, defer to the node's on-disk config. (Note: Eventually, flags will be deprecated and on-disk config will be the final fallback.)
+- else defer to config from flags
+- Finally, start the configuration sync loop in a background thread
+
+
+### Storing config locally (checkpointing)
+
+Initially, the Kubelet will store the most recent configmap it got from the API server internally, and compare that to new values it gets from the API server. It will restart when these configmaps differ in order to apply the new config.
+
+This is a rudimentary solution, and @aaronlevy brought up some good reasons to checkpoint the config, and store it externally from the Kubelet:
```

Comment:
Checkpointing is necessary for letting the kubelet run even with the api server is temporarily unreachable.

The fact that new configs might be incompatible with old configs is a problem that we should not try to optimize. In-compatible config changes might require a node reboot and that needs to be communicated to users as part of the release.
Adding checkpointing into this mix is too confusing in IMHO.